src/coprocessor/dag/batch/executors/slow_hash_aggr_executor.rs

```diff
@@ -188,63 +181,108 @@ impl<Src: BatchExecutor> AggregationExecutorImpl<Src> for SlowHashAggregationImp
+            &entities.each_aggr_exprs,
+            src_schema,
+            &mut input,
+        )?;
+        assert!(self.group_by_results.is_empty());
+        assert!(self.aggr_expr_results.is_empty());
+        unsafe {
+            eval_exprs_no_lifetime(
+                context,
+                &self.group_by_exps,
+                src_schema,
+                &input,
+                &mut self.group_by_results,
+            )?;
+            eval_exprs_no_lifetime(
+                context,
+                &entities.each_aggr_exprs,
+                src_schema,
+                &input,
+                &mut self.aggr_expr_results,
+            )?;
         }
 
-        for (group_key, group_key_offsets) in group_by_keys.into_iter().zip(group_by_keys_offsets) {
-            match self.groups.entry(group_key) {
+        for row_index in 0..rows_len {
+            let offset_begin = self.group_key_buffer.len();
+            for group_by_result in &self.group_by_results {
+                match group_by_result {
+                    RpnStackNode::Vector { value, field_type } => {
+                        value.encode(row_index, field_type, &mut self.group_key_buffer)?;
+                        self.group_key_offsets.push(self.group_key_buffer.len());
+                    }
+                    // we have checked that group by cannot be a scalar
+                    _ => unreachable!(),
+                }
+            }
+            let group_key_ref_unsafe = GroupKeyRefUnsafe {
+                buffer: (&*self.group_key_buffer).into(),
+                begin: offset_begin,
+                end: self.group_key_buffer.len(),
+            };
+            let group_len = self.groups.len();
+            let group_index = match self.groups.entry(group_key_ref_unsafe) {
                 Entry::Vacant(entry) => {
-                    let offset = self.states.len();
-                    self.states_offset_each_row.push(offset);
-                    entry.insert(GroupInfo {
-                        states_start_offset: offset,
-                        group_key_offsets,
-                    });
+                    // if it's a new group, the group index is the current group count
+                    entry.insert(group_len);
                     for aggr_fn in &entities.each_aggr_fn {
                         self.states.push(aggr_fn.create_state());
                     }
+                    group_len
                 }
                 Entry::Occupied(entry) => {
-                    self.states_offset_each_row
-                        .push(entry.get().states_start_offset);
+                    // remove the duplicated group key
+                    self.group_key_buffer.truncate(offset_begin);
+                    self.group_key_offsets
+                        .truncate(self.group_key_offsets.len() - group_by_len);
+                    *entry.get()
+                }
+            };
+            let state_begin = group_index * aggr_fn_len;
```

Comment:
I think you are still able to build a `states_offset_each_row`. Then `update_each_row_states_by_offset` can be reused.

Reply:
Changed to reuse `update_each_row_states_by_offset` now.